### ISSUE TITLE
Importing some resource types has always changes in plan and cannot be applied

### ISSUE THREAD

**Issue by minamijoyo** [minamijoyo/tfmigrate#4]
The current implementation of `tfmigrate apply` command requires that a plan has no changes for safe refactoring. However, importing some resource types has always changes in plan and cannot be applied.

For example, a `force_destroy` attribute in `aws_s3_bucket` is a provider implementation control parameter which cannot be imported from a real resource and will be computed default to `false`. In this case, it will change from `null` to `false` in the plan after import.

The first idea I came up with in my mind was to add an option like `tfmigrate apply -force`, but it's not suitable for GitOps because we always need to set `-force` in a CI/CD pipeline. Obviously, it's not safe to use.

At least, a `force` parameters should be set for each migration file as a `migration` block attribute. Furthermore, it would be great if we could declare which diffs could be ignored like `ignore_changes` attribute in Terraform configurations.

As a workaround, you can do it by splitting two phases:
1. Import a resource by temporarily adding `ignore_changes` to `*.tf` and `tfmigrate apply`
2. Remove `ignore_changes` from `*.tf` and `terraform apply`

I know it's not ideal but it should work.

**Comment by entropitor** [comment]
Maybe one way to help into doing this is add a command e.g. `tfmigrate export-bash` that outputs a bash script that you can run to actually run the migrations you have in your config file.

I'm not sure if this played before, but than the config file also doesn't have to be as close to the terraform commands and can be more like:

```hcl
migration "state" "test" {
  dir = "dir1"
  action {
    operation = "rm"
    address = "aws_security_group.baz"
  }
  action {
    operation = "import"
    address = "aws_security_group.qux"
    external_id = "qux"
  }
  action {
    operation = "mv"
    from = "aws_security_group.foo"
    to = "aws_security_group.foo2"
  }
}
```

**Comment by minamijoyo** [comment]
Simulating the current implementation in bash is not so easy because tfmigrate uses many temporarily files not to update remote state, and removes them on exit even if error occurs using defer mechanism in Go.

Anyway, I'm not sure why does exporting bash script solve the issue. If it solved the issue, the script would ignore changes in plan. That's not the same behavior in Go implementation. Whether exporting script or not, we need to control whether ignoring changes or not.

**Comment by entropitor** [comment]
The export script could be used to apply the changes. By running the migrations "manually" instead of letting tfmigrate do it. (In case there are actually still changes left, e.g. because of an import or because the refactor has side-effects that we want).

**Comment by minamijoyo** [comment]
If you want to run the script only when a migration has side-effect, I feel it probably have the same issue as `migrate apply -force` which I mentioned the issue description. It's not declarative when the diff should be ignored.

**Comment by syndicut** [comment]
One other use case for this feature will be when refactoring to modules cases come additional resources to be created and it's expected. If you wouldn't mind I can work on a PR adding "force" flag to migration file, like mentioned here:
> At least, a force parameters should be set for each migration file as a migration block attribute.

**Comment by minamijoyo** [comment]
@syndicut Thanks. A pull request is appreciate.

> One other use case for this feature will be when refactoring to modules cases come additional resources to be created and it's expected.

May I ask how it happens? I'm just curious.

**Comment by syndicut** [comment]
> May I ask how it happens? I'm just curious.

Suppose we have too almost identical environments (all in separate dirs, ommited for simplicity): https://gist.github.com/syndicut/b26f1bf2d6cb921d3c70b929fc27a852
We want to DRY them and it's ok to also have aws_security_group.baz in environment2
So we move security groups to module and include it:
https://gist.github.com/syndicut/efd1080ec3ea9959e87bdaad1d4284ad
Now we need to migrate states for env1 and env2, but we will always have aws_security_group.baz created in env2

We can do refactoring in two steps - add aws_security_group.baz to env2, apply these changes and then make refactoring to modules, but it can be a little bit tedious in real world examples with multiple environments and mutliple resources.

**Comment by syndicut** [comment]
@minamijoyo do you think it should be introduced in a new schema version for history file?

**Comment by minamijoyo** [comment]
@syndicut Thank you for sharing your use-case.

It's interesting. Ideally we want to be diligent, but it's not always true, sometimes be lazy.
How much we can be lazy, it depends 😉 

> do you think it should be introduced in a new schema version for history file?

No.
We don't need to record the fact that a migration applied force or not in history.
It's already defined in the migration file.

**Comment by minamijoyo** [comment]
Fixed by #10 in v0.2.1. 🚀 

It added the `force` option to a migration file. Although, it doesn't include the `ignore_chages` option, but I think the original issue was solved, so I'm going to close this issue. If you want to need the `ignore_changes` option, feel free to open a new issue.

Great thanks to @syndicut  🎉
